src/test/ui/consts/miri_unleashed/box.rs

```diff
@@ -1,13 +1,13 @@
 // compile-flags: -Zunleash-the-miri-inside-of-you
-#![feature(box_syntax)]
+#![feature(rustc_attrs, stmt_expr_attributes)]
 #![allow(const_err)]
 
 use std::mem::ManuallyDrop;
 
 fn main() {}
 
 static TEST_BAD: &mut i32 = {
-    &mut *(box 0)
+    &mut *(#[rustc_box] Box::new(0))
```

Comment:
This feels a little suspicious to me -- it looks like this test should be fine with just Box::new, without the rustc_box attr?

Reply:
That would change the error to:

    calling non-const function `Box::<i32>::new`

which is I think not what the test is for, it's specific to box syntax. I've removed it from the PR and defer it to later.

Reply:
Interesting. It seems like a bug in const checking to me that we allow "raw" allocation via rustc_box, but not via the user-facing function - seems like that could be an unintentional stability hole (e.g., when using vec![....] around this. Would probably be good to file an issue so we can dig deeper.

Reply:
Allocation isn't allowed in either instance:

* `box 0` has the current error: "calling non-const function `alloc::alloc::exchange_malloc`" (it's a compile-fail test after all)
* `#[rustc_box] Box::new(0)` lowers the call early on during AST -> HIR lowering, and thus the pass that does the checking treats it as `box 0` and gives it the same error, so it's not passing compilation. In the future, I plan to either remove `#[rustc_box]` [for an alternative](https://github.com/rust-lang/rust/issues/49733#issuecomment-1207075047), or lower it later, e.g. [during MIR building](https://github.com/rust-lang/rust/pull/97293#issuecomment-1144702190).
* `Box::new(0)` errors because the `new` function is not const.

I don't see a stability hole here.

It's just that (I presume) the test is here not to check what happens when you try to invoke a non-const function, because there are probably other tests for that. But instead the test is supposed to check what happens when you try to invoke `box` syntax to allocate. Maybe, if `box` syntax is removed, the `box` usage in the test should be replaced with `#[rustc_box]` (what I would prefer). Maybe it should be replaced with `Box::new`, but I would keep that discussion to a later PR.

Reply:
Ah, I see. Yes, that makes sense to me. OK to leave it for the future too.